spec.md

```diff
@@ -0,0 +1,100 @@
+For animation support, see [Animations](#animations).
+
+## Frame Structure
+A single frame is composed of multiple numerically-indexed tables, each representing a single line of the image.
+Additionally, an optional per-frame palette can be provided to alter the colors of that frame only. For more information on palettes, see [Palettes](#palettes).
+
+When the image is an animation, an optional `duration` field can also be provided. This field overrides the `secondsPerFrame` field set in the metadata, and therefore sets a specific duration for that frame only.
+
+![Frame Data](images/frame-structure.png)
+
+Each line is composed of three `string` values forming a blit table, that can be unpacked directly to `term.blit` when rendering the image.
+
+Spaces in any of the blit strings should make the corresponding color of the pixel transparent, according to the following rules:
+| Foreground | Background | Result |
+|:-:|:-:|---|
+| Colored | Colored | The pixel is rendered as is, using the colors set in the blit strings. No transparency is applied. |
+| Colored | Transparent | The background color of the pixel that is about to be replaced is used as the new background color. |
+| Transparent | Colored | The background color of the pixel that is about to be replaced is used as the new foreground color. |
+| Transparent | Transparent | The pixel is not rendered.
+
+## Animations
+bimg can support multiple frames per image file. Each frame is numerically indexed at the root of the table, along the metadata. This allows easy retrieval with `ipairs`.
+
+In the event multiple frames are provided, the image is considered as an animation. Consequently, the `animation` boolean should be set to `true`, and the `secondsPerFrame` field should be present.
+
+# Palettes
+Palettes allows the modification of the existing 16 ComputerCraft colors available in the `colors` API to custom RGB values, applied to the whole terminal at once.
+bimg allows the use of such palettes to render images more realistically, allowing for image effects (like dithering) to be applied.
+
+Palettes are formatted as a table, where the keys corresponds to the ComputerCraft color values, and the values the new RGB colors to apply.
+Values are contained themselves in a table, and can be either represented by a triplet of float numbers ranged from 0 to 1, each representing a color component, or a single integer number representing the three components at once:
+ - `{ 1, 1, 0 }` sets the red and green components to full, outputting a bright yellow.
+ - `{ 0xFFFF00 }` has the same effect, and will also provide a bright yellow.
+
+This allows both versions of `term.setPaletteColor()` to be supported through a single call, by unpacking the color table into the last argument:
+```lua
+for color, tbl in pairs(palette) do
+    term.setPaletteColor(color, table.unpack(tbl))
+end
+```
+
+Full example of a bimg color palette, featuring both color formats:
```

Comment:
Heads up that `colors` isn't available inside `textutils.unserialize`. I recommend using indexes 0-15 (or 1-16 if you really don't want to deal with `[0]` keys), then use `2^i` to calculate the final color:
```lua
palette = {
  [0] = {1, 0, 0},
  {0, 1, 0},
  {0x0000FF},
  [7] = {0.5, 0.5, 0.5}
}
-- parse:
for color, tbl in pairs(palette) do
  term.setPaletteColor(2^color, table.unpack(tbl))
end
```
This behavior will also match up nicely with the hex strings.

Reply:
This breaks compatibility with existing tools. I don't disagree with the suggestion, but existing tools will need to be updated to be compatible with this change

Reply:
afaik yours is the only existing tool.
The solution we had with this is a custom version of `textutils.unserialize` where it passed a custom `colours` table, where it only had the constants and not any of it's functions.

Reply:
Also, as a sidenote, program-generated ones will use the actual numbers (simply because making it into a table and serializing it causes the constant to be used or something (insert cool tech words here))